README.md
# Interrupt distributor

A unit to distribute N interrupt sources to M targets. The number of output
interrupt lines is N*M.

## Register map

TODO

## Testbench
In `tb/` there is a simple testbench. Run it by calling

```
make build
make sim
run -a
```


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [x.0.0] - 2023-05-03
### Added
- Repository initialization


rtl/irq_router.sv
// Copyright 2023 ETH Zurich and University of Bologna.
// Solderpad Hardware License, Version 0.51, see LICENSE for details.
// SPDX-License-Identifier: SHL-0.51
//
// Alessandro Ottaviano <aottaviano@iis.ee.ethz.ch>
// Robert Balas <balasr@iis.ee.ethz.ch>

// Distribute NumIntrSrc interrupts to NumIntrTargets targets according to a
// configurable mask. 1s in the mask indicate valid targets for the single
// interrupt line.

`include "register_interface/assign.svh"
`include "register_interface/typedef.svh"

module irq_router import irq_router_reg_pkg::*; #(
  parameter type         reg_req_t      = logic,
  parameter type         reg_rsp_t      = logic,
  parameter int unsigned NumIntrSrc     = 64,
  parameter int unsigned NumIntrTargets = 2
) (
  input logic                                  clk_i,
  input logic                                  rst_ni,

  input reg_req_t                              reg_req_i,
  output reg_rsp_t                             reg_rsp_o,

  input logic                 [NumIntrSrc-1:0]  irqs_i,
  output logic  [NumIntrTargets*NumIntrSrc-1:0] irqs_distributed_o
);

  if (NumIntrTargets > 32)
    $fatal(1, "maximum number supported targets is 32");

  irq_router_reg2hw_t [NumIntrSrc-1:0] irq_router_reg2hw;
  irq_router_hw2reg_t [NumIntrSrc-1:0] irq_router_hw2reg;

  reg_req_t [NumIntrSrc-1:0] reg_int_req;
  reg_rsp_t [NumIntrSrc-1:0] reg_int_rsp;

  logic [15:0] which_irq;

  // Top level address decoding and bux muxing. All registers of the irq
  // router are 32 bits wide.
  always_comb begin : irq_router_addr_decode
    which_irq = reg_req_i.addr[15:2];

    reg_int_req = '0;
    reg_rsp_o   = '0;

    reg_int_req[which_irq] = reg_req_i;
    reg_int_req[which_irq].addr = reg_req_i.addr[1:0];
    reg_rsp_o = reg_int_rsp[which_irq];
  end

  // Target configuration registers
  for (genvar i = 0; i < NumIntrSrc; i++) begin : gen_irq_router_regs
    irq_router_reg_top #(
      .reg_req_t(reg_req_t),
      .reg_rsp_t(reg_rsp_t)
    ) i_irq_router_reg_top (
      .clk_i,
      .rst_ni,

      .reg2hw (irq_router_reg2hw[i]),
      .hw2reg (irq_router_hw2reg[i]),

      .reg_req_i(reg_int_req[i]),
      .reg_rsp_o(reg_int_rsp[i]),

      .devmode_i(1'b1)
    );
  end // block: gen_irq_router_regs

  // Support array of NumIntrTargets*NumIntrSrc from input interrupt array
  logic [NumIntrTargets*NumIntrSrc-1:0] irqs_cloned;
  for (genvar i = 0; i < NumIntrTargets; i++) begin : gen_irq_il
    for (genvar j = 0; j < NumIntrSrc; j++) begin : gen_irq_jl
      assign irqs_cloned[j+(i*NumIntrSrc)] = irqs_i[j];
    end
  end

  // Array of masks for NumIntrTargets*NumIntrSrc
  logic [NumIntrTargets*NumIntrSrc-1:0] irqs_target_mask;
  for (genvar i = 0; i < NumIntrSrc; i++) begin : gen_mask_il
    for (genvar j = 0; j < NumIntrTargets; j++) begin : gen_mask_jl
      assign irqs_target_mask[i+(NumIntrSrc*j)] = irq_router_reg2hw[i].irq_target_mask.q[j];
    end
  end

  // Distribution of NumIntrSrc interrupts to NumIntrTargets targets according to mask
  assign irqs_distributed_o = irqs_cloned & irqs_target_mask;

endmodule // irq_router


rtl/irq_router_reg_pkg.sv
// Copyright lowRISC contributors.
// Licensed under the Apache License, Version 2.0, see LICENSE for details.
// SPDX-License-Identifier: Apache-2.0
//
// Register Package auto-generated by `reggen` containing data structure

package irq_router_reg_pkg;

  // Address widths within the block
  parameter int BlockAw = 2;

  ////////////////////////////
  // Typedefs for registers //
  ////////////////////////////

  typedef struct packed {
    logic [31:0] q;
  } irq_router_reg2hw_irq_target_mask_reg_t;

  typedef struct packed {
    logic [31:0] d;
    logic        de;
  } irq_router_hw2reg_irq_target_mask_reg_t;

  // Register -> HW type
  typedef struct packed {
    irq_router_reg2hw_irq_target_mask_reg_t irq_target_mask; // [31:0]
  } irq_router_reg2hw_t;

  // HW -> register type
  typedef struct packed {
    irq_router_hw2reg_irq_target_mask_reg_t irq_target_mask; // [32:0]
  } irq_router_hw2reg_t;

  // Register offsets
  parameter logic [BlockAw-1:0] IRQ_ROUTER_IRQ_TARGET_MASK_OFFSET = 2'h 0;

  // Register index
  typedef enum int {
    IRQ_ROUTER_IRQ_TARGET_MASK
  } irq_router_id_e;

  // Register width information to check illegal writes
  parameter logic [3:0] IRQ_ROUTER_PERMIT [1] = '{
    4'b 1111  // index[0] IRQ_ROUTER_IRQ_TARGET_MASK
  };

endpackage



rtl/irq_router_reg_top.sv
// Copyright lowRISC contributors.
// Licensed under the Apache License, Version 2.0, see LICENSE for details.
// SPDX-License-Identifier: Apache-2.0
//
// Register Top module auto-generated by `reggen`


`include "common_cells/assertions.svh"

module irq_router_reg_top #(
  parameter type reg_req_t = logic,
  parameter type reg_rsp_t = logic,
  parameter int AW = 2
) (
  input logic clk_i,
  input logic rst_ni,
  input  reg_req_t reg_req_i,
  output reg_rsp_t reg_rsp_o,
  // To HW
  output irq_router_reg_pkg::irq_router_reg2hw_t reg2hw, // Write
  input  irq_router_reg_pkg::irq_router_hw2reg_t hw2reg, // Read


  // Config
  input devmode_i // If 1, explicit error return for unmapped register access
);

  import irq_router_reg_pkg::* ;

  localparam int DW = 32;
  localparam int DBW = DW/8;                    // Byte Width

  // register signals
  logic           reg_we;
  logic           reg_re;
  logic [BlockAw-1:0]  reg_addr;
  logic [DW-1:0]  reg_wdata;
  logic [DBW-1:0] reg_be;
  logic [DW-1:0]  reg_rdata;
  logic           reg_error;

  logic          addrmiss, wr_err;

  logic [DW-1:0] reg_rdata_next;

  // Below register interface can be changed
  reg_req_t  reg_intf_req;
  reg_rsp_t  reg_intf_rsp;


  assign reg_intf_req = reg_req_i;
  assign reg_rsp_o = reg_intf_rsp;


  assign reg_we = reg_intf_req.valid & reg_intf_req.write;
  assign reg_re = reg_intf_req.valid & ~reg_intf_req.write;
  assign reg_addr = reg_intf_req.addr[BlockAw-1:0];
  assign reg_wdata = reg_intf_req.wdata;
  assign reg_be = reg_intf_req.wstrb;
  assign reg_intf_rsp.rdata = reg_rdata;
  assign reg_intf_rsp.error = reg_error;
  assign reg_intf_rsp.ready = 1'b1;

  assign reg_rdata = reg_rdata_next ;
  assign reg_error = (devmode_i & addrmiss) | wr_err;


  // Define SW related signals
  // Format: <reg>_<field>_{wd|we|qs}
  //        or <reg>_{wd|we|qs} if field == 1 or 0
  logic [31:0] irq_target_mask_qs;
  logic [31:0] irq_target_mask_wd;
  logic irq_target_mask_we;

  // Register instances
  // R[irq_target_mask]: V(False)

  prim_subreg #(
    .DW      (32),
    .SWACCESS("RW"),
    .RESVAL  (32'h1)
  ) u_irq_target_mask (
    .clk_i   (clk_i    ),
    .rst_ni  (rst_ni  ),

    // from register interface
    .we     (irq_target_mask_we),
    .wd     (irq_target_mask_wd),

    // from internal hardware
    .de     (hw2reg.irq_target_mask.de),
    .d      (hw2reg.irq_target_mask.d ),

    // to internal hardware
    .qe     (),
    .q      (reg2hw.irq_target_mask.q ),

    // to register interface (read)
    .qs     (irq_target_mask_qs)
  );




  logic [0:0] addr_hit;
  always_comb begin
    addr_hit = '0;
    addr_hit[0] = (reg_addr == IRQ_ROUTER_IRQ_TARGET_MASK_OFFSET);
  end

  assign addrmiss = (reg_re || reg_we) ? ~|addr_hit : 1'b0 ;

  // Check sub-word write is permitted
  always_comb begin
    wr_err = (reg_we &
              ((addr_hit[0] & (|(IRQ_ROUTER_PERMIT[0] & ~reg_be)))));
  end

  assign irq_target_mask_we = addr_hit[0] & reg_we & !reg_error;
  assign irq_target_mask_wd = reg_wdata[31:0];

  // Read data return
  always_comb begin
    reg_rdata_next = '0;
    unique case (1'b1)
      addr_hit[0]: begin
        reg_rdata_next[31:0] = irq_target_mask_qs;
      end

      default: begin
        reg_rdata_next = '1;
      end
    endcase
  end

  // Unused signal tieoff

  // wdata / byte enable are not always fully used
  // add a blanket unused statement to handle lint waivers
  logic unused_wdata;
  logic unused_be;
  assign unused_wdata = ^reg_wdata;
  assign unused_be = ^reg_be;

  // Assertions for Register Interface
  `ASSERT(en2addrHit, (reg_we || reg_re) |-> $onehot0(addr_hit))

endmodule

module irq_router_reg_top_intf
#(
  parameter int AW = 2,
  localparam int DW = 32
) (
  input logic clk_i,
  input logic rst_ni,
  REG_BUS.in  regbus_slave,
  // To HW
  output irq_router_reg_pkg::irq_router_reg2hw_t reg2hw, // Write
  input  irq_router_reg_pkg::irq_router_hw2reg_t hw2reg, // Read
  // Config
  input devmode_i // If 1, explicit error return for unmapped register access
);
 localparam int unsigned STRB_WIDTH = DW/8;

`include "register_interface/typedef.svh"
`include "register_interface/assign.svh"

  // Define structs for reg_bus
  typedef logic [AW-1:0] addr_t;
  typedef logic [DW-1:0] data_t;
  typedef logic [STRB_WIDTH-1:0] strb_t;
  `REG_BUS_TYPEDEF_ALL(reg_bus, addr_t, data_t, strb_t)

  reg_bus_req_t s_reg_req;
  reg_bus_rsp_t s_reg_rsp;
  
  // Assign SV interface to structs
  `REG_BUS_ASSIGN_TO_REQ(s_reg_req, regbus_slave)
  `REG_BUS_ASSIGN_FROM_RSP(regbus_slave, s_reg_rsp)

  

  irq_router_reg_top #(
    .reg_req_t(reg_bus_req_t),
    .reg_rsp_t(reg_bus_rsp_t),
    .AW(AW)
  ) i_regs (
    .clk_i,
    .rst_ni,
    .reg_req_i(s_reg_req),
    .reg_rsp_o(s_reg_rsp),
    .reg2hw, // Write
    .hw2reg, // Read
    .devmode_i
  );
  
endmodule




tb/irq_router_tb.sv
// Copyright 2022 ETH Zurich and University of Bologna.
// Copyright and related rights are licensed under the Solderpad Hardware
// License, Version 0.51 (the "License"); you may not use this file except in
// compliance with the License.  You may obtain a copy of the License at
// http://solderpad.org/licenses/SHL-0.51. Unless required by applicable law
// or agreed to in writing, software, hardware and materials distributed under
// this License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
// CONDITIONS OF ANY KIND, either express or implied. See the License for the
// specific language governing permissions and limitations under the License.

// Wolfgang Roenninger <wroennin@iis.ee.ethz.ch>
// Maicol Ciani <maicol.ciani@unibo.it>
// Robert Balas <balasr@iis.ee.ethz.ch>
// Alessandro Ottaviano <aottaviano@iis.ee.ethz.ch>

// irq router unit basic testbench

`include "axi/assign.svh"
`include "axi/typedef.svh"
`include "register_interface/assign.svh"
`include "register_interface/typedef.svh"

module irq_router_tb import axi_pkg::*; #(
  parameter int NumIntrSrc = 16,
  parameter int NumIntrTargets = 8
)(
);

  // timing parameters
  localparam time CyclTime = 10ns;
  localparam time ApplTime =  2ns;
  localparam time TestTime =  8ns;

  // axi configuration
  localparam int unsigned AxiAddrWidth      =  32'd32;    // Axi Address Width
  localparam int unsigned AxiDataWidth      =  32'd32;    // Axi Data Width

  typedef logic [AxiAddrWidth-1:0] addr_t;
  typedef logic [AxiDataWidth-1:0] data_t;
  typedef logic [AxiDataWidth/8-1:0] strb_t;

  typedef enum addr_t {
    IRQ_TARGET_MASK = addr_t'(0 * AxiDataWidth/8)
  } reg_addr_e;

  typedef axi_test::axi_lite_rand_master #(
    // AXI interface parameters
    .AW ( AxiAddrWidth        ),
    .DW ( AxiDataWidth        ),
    // Stimuli application and test time
    .TA ( ApplTime            ),
    .TT ( TestTime            ),
    .MIN_ADDR ( 32'h0000_0000 ),
    .MAX_ADDR ( 32'h0001_3000 ),
    .MAX_READ_TXNS  ( 10 ),
    .MAX_WRITE_TXNS ( 10 )
  ) rand_lite_master_t;

  // DUT signals
  logic               clk;
  logic               rst_n;
  logic               end_of_sim;
  logic [NumIntrSrc-1:0] irqs = '0;
  logic [NumIntrSrc*NumIntrTargets-1:0] irqs_distributed;
  int unsigned                          mask = 0;

  int unsigned test_failed;

  // AXI Interfaces
  AXI_LITE #(
    .AXI_ADDR_WIDTH ( AxiAddrWidth      ),
    .AXI_DATA_WIDTH ( AxiDataWidth      )
  ) master ();
  AXI_LITE_DV #(
    .AXI_ADDR_WIDTH ( AxiAddrWidth      ),
    .AXI_DATA_WIDTH ( AxiDataWidth      )
  ) master_dv (clk);

  `AXI_LITE_ASSIGN(master, master_dv)

  `AXI_LITE_TYPEDEF_AW_CHAN_T(aw_chan_lite_t, addr_t)
  `AXI_LITE_TYPEDEF_W_CHAN_T(w_chan_lite_t, data_t, strb_t)
  `AXI_LITE_TYPEDEF_B_CHAN_T(b_chan_lite_t)
  `AXI_LITE_TYPEDEF_AR_CHAN_T(ar_chan_lite_t, addr_t)
  `AXI_LITE_TYPEDEF_R_CHAN_T(r_chan_lite_t, data_t)
  `AXI_LITE_TYPEDEF_REQ_T(axi_lite_req_t, aw_chan_lite_t, w_chan_lite_t, ar_chan_lite_t)
  `AXI_LITE_TYPEDEF_RESP_T(axi_lite_resp_t, b_chan_lite_t, r_chan_lite_t)

  axi_lite_req_t axi_lite_req;
  axi_lite_resp_t axi_lite_resp;

  `AXI_LITE_ASSIGN_TO_REQ(axi_lite_req, master)
  `AXI_LITE_ASSIGN_FROM_RESP(master, axi_lite_resp)

  // AXI Lite to reg
  `REG_BUS_TYPEDEF_REQ(reg_req_t, addr_t, data_t, strb_t)
  `REG_BUS_TYPEDEF_RSP(reg_rsp_t, data_t)

  reg_req_t reg_req;
  reg_rsp_t reg_rsp;

  axi_lite_to_reg #(
    .ADDR_WIDTH(AxiAddrWidth),
    .DATA_WIDTH(AxiDataWidth),
    .BUFFER_DEPTH(1),
    .DECOUPLE_W(0),
    .axi_lite_req_t(axi_lite_req_t),
    .axi_lite_rsp_t(axi_lite_resp_t),
    .reg_req_t(reg_req_t),
    .reg_rsp_t(reg_rsp_t)
  ) i_axi_lite_to_reg (
    .clk_i (clk),
    .rst_ni(rst_n),
    .axi_lite_req_i (axi_lite_req),
    .axi_lite_rsp_o (axi_lite_resp),
    .reg_req_o(reg_req),
    .reg_rsp_i(reg_rsp)
  );

  // DUT
  irq_router #(
      .reg_req_t      (reg_req_t),
      .reg_rsp_t      (reg_rsp_t),
      .NumIntrSrc     (NumIntrSrc),
      .NumIntrTargets (NumIntrTargets)
  ) i_irq_router (
      .clk_i     (clk),
      .rst_ni    (rst_n),
      .reg_req_i (reg_req),
      .reg_rsp_o (reg_rsp),
      .irqs_i    (irqs),
      .irqs_distributed_o (irqs_distributed)
  );

  clk_rst_gen #(
    .ClkPeriod    ( CyclTime ),
    .RstClkCycles ( 5        )
  ) i_clk_gen (
    .clk_o (clk),
    .rst_no(rst_n)
  );

`define TEST_READ_REG(ADDR, EXPECTED) \
    $display(`"%0t MST_0> Read register ``ADDR `", $time()); \
    lite_axi_master.read(``ADDR, axi_pkg::prot_t'('0), data, resp); \
    assert (data == data_t'(``EXPECTED)) else begin test_failed++; $error("Unexpected result"); end \
    assert (resp == axi_pkg::RESP_OKAY) else begin test_failed++; $error("Unexpected result"); end

`define TEST_WRITE_READ_REG(ADDR, EXPECTED) \
    $display(`"%0t MST_0> write/read register ``ADDR `", $time()); \
    lite_axi_master.write(``ADDR, axi_pkg::prot_t'('0), data_t'(``EXPECTED), 4'hf, resp);  \
    assert (resp == axi_pkg::RESP_OKAY) else begin test_failed++; $error("Unexpected result"); end \
    lite_axi_master.read(``ADDR, axi_pkg::prot_t'('0), data, resp); \
    assert (data == data_t'(``EXPECTED)) else begin test_failed++; $error("Unexpected result"); end \
    assert (resp == axi_pkg::RESP_OKAY) else begin test_failed++; $error("Unexpected result"); end

  initial begin : proc_master
    automatic rand_lite_master_t lite_axi_master = new ( master_dv, "MST_0");
    automatic data_t          data = '0;
    automatic axi_pkg::resp_t resp = axi_pkg::RESP_SLVERR;
    automatic addr_t          irq_router_base = '0;
    end_of_sim <= 1'b0;
    lite_axi_master.reset();
    @(posedge rst_n);

    // generate targets for each interrupt
    for (int k = 0; k < NumIntrSrc; k++ ) begin : gen_targets

      irq_router_base = 4 * k; // each programmable target occupies 4 bytes
      // Read all registers and compare their results
      $info("Initial test by reading each register");
      `TEST_READ_REG(irq_router_base + IRQ_TARGET_MASK, 32'h0);

      // write read scratchpad registers
      $info("write/read tests");
      // per-interrupt mask
      mask = $urandom_range(0, 2**NumIntrTargets);
      `TEST_WRITE_READ_REG(irq_router_base + IRQ_TARGET_MASK, mask);

    end // block: gen_targets

    // randomize input interrupts. Assume level-sensitive interrupts
    irqs = $urandom_range(0, 2**NumIntrSrc);

    end_of_sim <= 1'b1;
  end

  initial begin : proc_stop_sim
    wait (end_of_sim);
    repeat (50) @(posedge clk);
    $display("Number of failed tests: %0d", test_failed);
    if (test_failed > 0) begin
        $fatal(1, "Assertion errors. Failure!");
    end else begin
        $info("Success.",);
    end
    $stop();
  end

endmodule
